src/accumulator/stump.rs

```diff
@@ -0,0 +1,139 @@
+use super::types;
+
+#[derive(Debug, Clone)]
+pub struct Stump {
+  leafs: u64,
+  roots: Vec<bitcoin_hashes::sha256::Hash>
+}
+
+impl Stump {
+  /// Creates an empty Stump
+  pub fn new() -> Self {
+    Stump {
+      leafs: 0,
+      roots: Vec::new()
+    }
+  }
+  /// Modify is the external API to change the accumulator state. Since order
+  /// matters, you can only modify, providing a list of utxos to be added, 
+  /// and txos (@TODO) to be removed, along with it's proof. Either may be
+  /// empty.
+  ///# Example
+  /// ```
+  ///   use rustreexo::accumulator::stump::Stump;
+  ///   let mut s = Stump::new();
+  ///   let utxos = vec![];
+  ///   let stxos = vec![];
+  ///   s.modify(&utxos, &stxos);
+  /// ```
+  pub fn modify(&mut self, utxos: &Vec<bitcoin_hashes::sha256::Hash>, _stxos: &Vec<bitcoin_hashes::sha256::Hash>) {
+    //remove
+    self.add(utxos);
+  }
+
+  /// Rewinds old tree state, this should be used in case of reorgs.
+  /// Takes the ownership over `old_state`.
+  ///# Example
+  /// ```
+  ///   use rustreexo::accumulator::stump::Stump;
+  ///   let mut s_old = Stump::new();
+  ///   let mut s_new = Stump::new();
+  ///   
+  ///   s_old.modify(&vec![], &vec![]);
+  ///   s_new = s_old.clone();
+  ///   s_new.modify(&vec![], &vec![]);
+  ///   
+  ///   // A reorg happened
+  ///   
+  ///   s_new.undo(s_old);  
+  ///```
+  pub fn undo(&mut self, old_state: Stump) {
+    self.leafs = old_state.leafs;
+    self.roots = old_state.roots;
+  }
+
+  /// Adds new leafs into the root
+  fn add(&mut self, utxos: &Vec<bitcoin_hashes::sha256::Hash>) {
+    for i in utxos.iter() {
+      self.add_single(*i);
+    }
+  }
+
+  fn add_single(&mut self, node: bitcoin_hashes::sha256::Hash) {
+    let mut h = 0;
+    // Iterates over roots, if we find a root that is not empty, we concatenate with
+    // the one we are adding and create new root, leaving this position empty. Stops
+    // when find an empty root.
+
+    // You can say if a root is empty, by looking a the binary representations of the
+    // number of leafs. If the h'th bit is one, then this position is occupied, empty 
+    // otherwise.
+    let mut to_add = node;
+    while (self.leafs >> h) & 1 == 1 {
+      let root = self.roots.pop();
+      if let Some(root) = root {
+        to_add = types::parent_hash(&root, &to_add);
+      }
+      h += 1;
+    }
+
+    self.roots.push(to_add);
+
+    self.leafs += 1;
+  }
+}
+
+
+
+#[cfg(test)]
+mod test {
+  use std::vec;
+  use bitcoin_hashes::{sha256, Hash, HashEngine};
+  use super::Stump;
+
+  
+  #[test]
```

Comment:
We should add more unit tests here to ensure the correctness of the code. Perhaps run the tests as macros (something like this https://github.com/BurntSushi/fst/blob/master/src/raw/tests.rs#L111-L158)

Reply:
Don't stress too much on this for now, I'll also write up some tests here.

Reply:
Just took a few minutes to try that out, I've got a macro (for now called `make_test_case`) which creates tests for any tree with just the elements and roots.  
Here's how you implement the provided tree.
```rust
    make_test_case!(            
        tree_with_6_elements,
        [0, 1, 2, 3, 4, 5, 6],
        [
            "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386",
            "9eec588c41d87b16b0ee226cb38da3864f9537632321d8be855a73d5616dcc73",
            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6"
        ]
    );
```
This one creates a new test function for each scenario. It is possible to have just one function with all cases within it, in that case, the macros would look like this:
```rust
    make_test_case!(
        [0, 1, 2, 3, 4, 5, 6],
        [
            "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386",
            "9eec588c41d87b16b0ee226cb38da3864f9537632321d8be855a73d5616dcc73",
            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6"
        ]
    );
```
Another thing that came to my mind is providing test cases as a json or xml, so the code just read those cases and test.

Reply:
>make_test_case!(
        [0, 1, 2, 3, 4, 5, 6],
        [
            "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386",
            "9eec588c41d87b16b0ee226cb38da3864f9537632321d8be855a73d5616dcc73",
            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6"
        ]
    );

This looks great :)

> Another thing that came to my mind is providing test cases as a json or xml, so the code just read those cases and test.

Yeah that's a good idea. We'd eventually want some test cases across different implementations so making some test cases in json would be good. Bitcoin Core has them https://github.com/bitcoin/bitcoin/blob/master/src/test/data/tx_invalid.json